File: matilda/commands/protobuf/list_converter.py
from typing import TypeVar, List

from google.protobuf.any_pb2 import Any

from matilda.commands.protobuf.protobuf_converter import ProtobufConverter
from matilda.generated.proto.some_pb2 import Some

T = TypeVar('T')


class ListConverter(ProtobufConverter[List[T]]):
    def __init__(self, inner_converter: ProtobufConverter[T]):
        self.__inner_converter = inner_converter

    def to_protobuf(self, values: List[T]) -> Some:
        return Some(any=map(self.__to_any, values))

    def __to_any(self, value: T) -> Any:
        any_message = Any()
        any_message.Pack(msg=self.__inner_converter.to_protobuf(value))
        return any_message

    def from_protobuf(self, message: Any) -> List[T]:
        some = Some()
        message.Unpack(some)
        return list(map(self.__inner_converter.from_protobuf, some.any))

File: agent/commands-generator-api/src/main/java/org/matilda/commands/types/ListConverter.kt
package org.matilda.commands.types

import pbandk.wkt.Any
import org.matilda.commands.protobuf.Some
import pbandk.pack
import pbandk.unpack

class ListConverter<T>(private val mInternalConverter: ProtobufConverter<T>) : ProtobufConverter<List<T>> {
    override fun convertToProtobuf(obj: List<T>): Some {
        return Some(obj.map { element: T ->
            Any.pack(mInternalConverter.convertToProtobuf(element))
        })
    }

    override fun convertFromProtobuf(obj: Any): List<T> {
        val some = obj.unpack(Some.Companion)
        return some.any.map {
            mInternalConverter.convertFromProtobuf(it)
        }
    }
}
